### DISCUSSION TITLE
What's next for Node.js Website?

### DISCUSSION THREAD

**Discussion by ovflowd** [nodejs/nodejs.org#5131] [Announcements]
<details>
  <summary>Premise (Outdated)</summary>
  
	Much work through these last years was done on the https://github.com/nodejs/nodejs.dev repository. From drafting the new design from [this](https://www.figma.com/file/lOxAGGg5KXb6nwie7zXkz6/NJ---Design-System) Figma, to implementation of the pages, curation of what should be the content of the new website, and much more. 
	
	These last months have been incredibly important; with fascinating work being pushed through the edge of this open-source furnace, we finally finished the main goals and features of the Website Redesign Initiative.
	
	This allowed us to focus on migrating the Node.js Website (this very own repository) [into a modern Framework](https://github.com/nodejs/nodejs.org/issues/4958) that would also be compatible with the work we've done on the `nodejs.dev` repository. These changes were the foundational milestone allowing us to adopt the content, design and feature set from https://nodejs.dev.
	
	But what's next? This discussion aims to highlight what the Website Team is coordinating to get done and how we want to make these changes transparent and allow the inner and outer community to contribute, As one of our core goals is to improve the Developer Experience and allow more people to be able to contribute towards the Node.js Website.
  
</details>

# Node.js Website Redesign

#### A quick preamble

Hey y'all 👋 we're currently in the process of redesigning the Node.js Website. This multi-year effort has evolved from many previous efforts (for the curious ones: The Node.js Website Redesign Initiative, The `nodejs.dev` repository, and others).

This process has brought a lot of technical refactoring, restructuring of the Website Repository, and change of technologies used. This has been a long process, but to keep the new contributors short, most of the historical information is available on the hidden comments, and all the current relevant information is available below and on the non-hidden comments).

## The Effort

The Node.js Website Redesign is the initiative of revamping the current design of the Node.js Website into the newly done [Figma Design](https://www.figma.com/file/pu1vZPqNIM7BePd6W8APA5/Node.js).

#### The process consists of 3 main steps: (Creating Components -> Creating Layouts -> Migrating Pages)

Note that Issues regarding the Website Redesign will have a `website redesign` label. Issues that are ready to be worked on will have a `Ready` status.

- Creating React Components that follow the designs and implementation specs from the Figmas.
  - Each Component is described within the Figma on the "Design System" Tab
  - Issues will be created for each Component by the Website Team Maintainers
    - Changes should not be done before these issues get created
    - Issues will describe how the Components should work and the validation process
    - Issues will describe where styles, hooks and other pieces of code should live and any other 3rd party needed information
  - Contributors and Newcomers are free to work on the implementation of these Issues
- Creating the Layouts that will be used for the Pages
  - Layouts will be done following the same process mentioned before
- Pages will be migrated from the [`nodejs.dev/content/`](https://github.com/nodejs/nodejs.dev/tree/main/content) folder into `pages/new` folder
  - The `pages/new` folder is a temporary folder
  - Instructions on how to migrate each page, such as:
    - Filenaming and Extension of each file
    - What should be changed on each migrated page [Markdown Frontmatter](https://daily-dev-tips.com/posts/what-exactly-is-frontmatter/)
    - Which Layout each migrated page should adopt
    - Which pages should not be migrated

#### Then, after this process is done and validated:

- The old pages, styles, components and layouts will be removed
- All pages under `pages/new` will be moved to `pages/`
- Other minor changes such as on `pages/_app.tsx` and `pages/_document.tsx` will be done

Most of the communication of each step will be done before we move to each step, so don't worry!

### Recommended Reading Material

We recommend reading the Contributing and Collaborator Guidelines and the comments below.

---

cc @nodejs/website

**Comment by ovflowd** [comment]
Also, props to @AugustinMauroy for coming up with this partial table of how much complexity it would be to port specific components. https://gist.github.com/AugustinMauroy/4dfa2b6eab2a00356e5fb75268363800

**Reply by AugustinMauroy** [reply]
About that all opinion / idea ... are welcome

**Reply by AugustinMauroy** [reply]
There are also things to discuss but I will make an issue/discussion at the appropriate time.

**Comment by manishprivet** [comment]
The port seems like big task, but definitely doable. However, IMO it would be great to have nodejs.dev website mapped to this `major/website-redesign` branch for more transparency on the changes that will be incoming.

**Reply by AugustinMauroy** [reply]
This is the list of React components, React Hooks, content and other about nodejs.dev
 https://gist.github.com/AugustinMauroy/4dfa2b6eab2a00356e5fb75268363800
 I hope it will help you and other people to understand the migration.
 
 > Don't forget that this is react but nodejs.dev is created with Gatsby and nodejs.org with Next.js

**Reply by ovflowd** [reply]
@AugustinMauroy I think @manishprivet knows enough the framework differences of both repositories as he was part of the original Website Redesign team :)

**Reply by manishprivet** [reply]
Yeah, I have also been advocating NextJS for a while, and was really glad seeing it in fruition! All thanks to @ovflowd 

Apologies for asking this, as I have been out of the loop for a few months, but are we still keeping the /learn page? The doc @AugustinMauroy shared does mention that.

**Reply by AugustinMauroy** [reply]
This should be a misunderstanding. The learn section is keep.
Would you mind giving me the part that led you to this confusion so I can update it.

**Reply by ovflowd** [reply]
@AugustinMauroy there's no confusion, he was just double-checking if the learn pages are going to be migrated, he didn't imply that your doc somewhere mentioned they wouldn't.

And yes, @manishprivet we will also bring the learn pages :)

**Comment by nazarepiedady** [comment]
I am following up, you are doing a good job.
When I have some free time I will try to help in what I can.

**Reply by ovflowd** [reply]
Thanks @nazarepiedady :)

**Comment by manishprivet** [comment]
Hi, I believe the list of things that need to be done here are going to be tracked through individual issues. Do we have an idea on when it's going to get started, and who is going to create the first set of issues? I wanted to jump in and contribute but I can't seem to find any timeline on this. Apologies in advance if I missed anything

**Reply by ovflowd** [reply]
Hey Manish! I'm going to start creating the issues tomorrow! Including how the directory structure probably should be, the first components we want to move, etc. 🙇

**Reply by manishprivet** [reply]
Great, and thanks! Excited to get this going

**Reply by ovflowd** [reply]
No worries. ALSO SUPER EXCITED LETS GO! (Where's @AugustinMauroy to say LETS GO TOO)

**Reply by AugustinMauroy** [reply]
Me...
LETS GOOOOOOO !
don't miss you have [this docs](https://gist.github.com/AugustinMauroy/4dfa2b6eab2a00356e5fb75268363800)

**Reply by manishprivet** [reply]
Can we add this doc in this repository? Maybe an issue can work which can be referenced in every subsequent issue?

**Reply by AugustinMauroy** [reply]
Good question. Personally I'm not in favor of an issue for this doc.
Friendly ping @ovflowd 👋

**Reply by ovflowd** [reply]
We don't need to make an issue for these items on the Gist, I will anyways create issues for each part, including the components and etc. I apologise for the delays 😅

**Comment by ovflowd** [comment]
I made a few issues. For now that's enough for what we have to work with. Going to make more as we progress.

**Comment by ovflowd** [comment]
Just an update/FYI we're already working on the Component migrations. You can see all the issues that are being worked on by filtering issues with the `website redesign` label.

**Comment by ovflowd** [comment]
**More updates:**

- Most of the Components are done already; some require/might need some refinement/polishment
- I've just wrote down https://github.com/nodejs/nodejs.org/issues/5403 this issue with details about further parts of the website redesign
- I'm working on the final touches to get the new website ready to some test deployments for Vercel

**Comment by ovflowd** [comment]
**More updates:**

- https://github.com/nodejs/build/issues/3366 was created containing the next steps for setting up the test environment (aka https://vercel.nodejs.org) for the split traffic between the Node.js website on Vercel and the distribution assets on our DO infrastructure.
- There's a small blocker in terms of i18n/routing that is an open question for Vercel right now (https://github.com/nodejs/nodejs.org/issues/5404) and I'm more than happy to get some support from people that have more solid experience with Next.js
  - The gist of this issue is that we want to be able that pages that are 404 in different languages can be detected if they exist in the default language or not.
  - After many hours of work we were finally able to come up with a solution [here](https://github.com/nodejs/nodejs.org/pull/5426)
- We're going to start migrating the remaining more complex components, such as Navigation and others.

**Comment by ovflowd** [comment]
**More updates:**

- We've merged our Dynamic Rendering implementation through [this PR](https://github.com/nodejs/nodejs.org/pull/5426)
  - We've further made some adjustments and bug fixes, and improvements with the following PRs:
    - https://github.com/nodejs/nodejs.org/pull/5467
    - https://github.com/nodejs/nodejs.org/pull/5474
    - https://github.com/nodejs/nodejs.org/pull/5469
    - https://github.com/nodejs/nodejs.org/pull/5475
- We're still migrating components but have a steady speed on migrating Components.
- @bmuenzenmeyer is doing some benchmarking that allows us to compare previous builds

**What's Next:**

- Continuing the Migration of Components
- Investigate Monorepo Solution for the Website (https://github.com/nodejs/nodejs.org/issues/5427)
- Work into getting the current Next.js Website on Vercel to replace https://nodejs.org as from (https://github.com/nodejs/build/issues/3366)

**Comment by bmuenzenmeyer** [comment]
from @ovflowd via Slack

Hey folks :wave: pretty important announcement here. We should put the migration of the components into a halt. The main reason is that we’re (The Foundation) are coming up with completely new designs for the Node.js website, and we want to put the current migration of components into a halt.
Mostly because we don’t know which components will become redundant and which ones will not. A lot of things will drift from the initial [nodejs.dev](http://nodejs.dev/) design.

- If a component on the new design does not exist but we already migrated, we will probably not delete them (only if they’re corner case components) as we might be able to repurpose them, but inevitably, some might be deleted.
- If a component that we already migrated exists in the new design, we will simply update its style and feature to match requirements
- The new design simplifies a lot of the Node.js website design and reduces the amount of components we need
- The new design Figma’s will be shared in the future once we’re closer to a final version
- The design will first get an “approval” from the TSC (more like getting their opinions and checking if they’re fine with it)

**Comment by ovflowd** [comment]
## :information_source: Updates on Development

Hey everyone! :wave: I'm excited to share some updates about our upcoming website redesign. Here's what you need to know:

### Design Progress :art:

We're getting closer to finalizing the new designs! Next Monday, we'll have pre-final reviews at the Foundation. The Figma Link to the new designs is available for internal use. Remember not to share it until it's ready for public release.

Please refrain from extensive design critiques as our focus is on the technical aspects. Your feedback is valuable, but let's allow the experts to lead in design decisions.

### Component Work :computer:

@bmuenzenmeyer and I will soon create/update issues for component work. Some pages might not be ready immediately, like API Docs, but we can still work on components. Note that not all designs will be incorporated, so these are not final.

### Design System and Icons :art: :symbols:

We're adopting Tailwind CSS as our base design system, offering consistent styles across components. Check out the [Tailwind Configuration](https://tailwindcss.com/docs/customizing-colors#using-custom-colors) for color and font variables. Heroicons will replace React Icons.

### Component Transition :arrows_counterclockwise:

Existing components will be remapped or replaced to align with the new design. This plan will be shared later this month, specifying deletions, updates, and new additions. We're discarding previous styles from nodejs.dev.

### Styling Components :nail_care:

Refer to the Figma Design System for component styles. You can often copy CSS directly. Minimize duplicated styles to maintain consistency.

### Code Editors :keyboard:

We're switching from PrismJS to Shiki with Rehype Pretty Code. Shiki is used by VS Code. [Themes](https://unpkg.com/browse/shiki@0.14.2/themes/github-light.json) for GitHub Light and Nord will be integrated.

### API Docs :clipboard:

- API Docs are separate and not part of the initial development.
- Some component designs are finalized and can be implemented.
- The overall layout of the API Docs will change.
- The goal is to have a self-hosted API Docs application on nodejs/node.
- API Components from nodejs/nodejs.org will be reused in the API Docs codebase.

### Other Updates :information_source:

- We're adopting Vercel OG Image Generation for blog images.
- Unnecessary packages like PrismJS and React Icons will be removed.
- Collaborator Guides and README will be updated.

### Development Progress :construction:

- As Figma designs approach finalization, development on new components is getting closer.
- You can start working on tasks like Tailwind setup and more.
- Component designs are mostly locked in, allowing us to begin implementing a few components.

Let's work together to bring these exciting changes to life! :rocket:

**Reply by ovflowd** [reply]
* The New Designs (Figma) are available here: https://www.figma.com/file/pu1vZPqNIM7BePd6W8APA5/Node.js
* The full content of this update is available through Slack: https://openjs-foundation.slack.com/archives/CVAMEJ4UV/p1691420914457079
  * The version above is a summary generated by ChatGPT.

**Reply by ghost** [reply]
@ovflowd , when are you planning to release the new version?

**Reply by ovflowd** [reply]
What do you mean by that?

**Reply by ghost** [reply]
@ovflowd , I mean seeing nodejs.org website as on those Figma designs. I wonder how long it would take?

**Reply by ovflowd** [reply]
I do not think I can evaluate the amount of time that this will take to reach completion. This is an open source project, not a product. It will solely depend on the free will and time of the contributors.

Being reasonable this will definitely take months.

**Comment by ovflowd** [comment]
## ℹ️ Updates on Development

The current Figma designs were updated to a semi-final version. All the Components on the [Component Design Syetem](https://www.figma.com/file/pu1vZPqNIM7BePd6W8APA5/Node.js?type=design&node-id=337-7756&mode=design&t=r5npkZd2OT2dveC8-0) are pretty much final and can be worked on.

We're now validating the structure and layouts of the Website Redesign with the Node.js TSC (Technical Steering Committee)

And we're creating issues pertinent to the Development of the new Website. You can already stay tuned on the Issues tab of this Repository to follow-up all the needed work :)

**Reply by bmuenzenmeyer** [reply]
Someone asked how to validate values across Figma and the finished code. I wanted to present a brief guide here, and would not be surprised if someone showed me a better way then 😄 

<details>
  <summary><H2>How to verify changes with Figma</H2></summary>
  
### 1 Find the thing you are building
	
In this case, I care about the color swatches. I found them on the Design System Page
	
![image](https://github.com/nodejs/nodejs.org/assets/298435/3ebab784-219e-4fd8-97cf-508cbb56440b)
	
If I click on the whole component, I might get a grouping, which doesn't show me any CSS property values.
	
<details><summary>example of a single click grouping</summary>
	
![image](https://github.com/nodejs/nodejs.org/assets/298435/76376961-468a-41a7-ad95-d2737208c543)
	
</details>
	
### 2 Double-click on the item
	
Select a single component by double clicking.
	
![image](https://github.com/nodejs/nodejs.org/assets/298435/b0b542f0-f76d-46a3-93fb-012b4dcef4c6)
	
Doing this will also reveal values on the right hand, properties tab
	
![image](https://github.com/nodejs/nodejs.org/assets/298435/95d92458-0789-463d-92e7-84a996197eb1)
	
### 3 Find the value you are looking for
	
Green 400? Ah, there is it. `#84BA64`
	
![image](https://github.com/nodejs/nodejs.org/assets/298435/ec0da0a7-3311-4a63-afc1-06bf1338336e)
	
### 4 Dev Mode?

You might have also noticed something called `Dev Mode` on the FIgma or those screenshots. Turn it on and off with this toggle.

![image](https://github.com/nodejs/nodejs.org/assets/298435/554e4955-9bc2-4c8b-be6e-ab1d18e762b8)
	
This gets more interesting when you look at a CSS component that is intended for direct usage on the screen, unlike the swatch. 
	
Let's look at some Text Styles
	
![image](https://github.com/nodejs/nodejs.org/assets/298435/6205f9fb-b8b8-43dc-8cb5-ccb92b5ee017)
	
Or a button
	
![image](https://github.com/nodejs/nodejs.org/assets/298435/9d471ded-2532-4418-9e71-86bc3a855e5d)
  
</details>

**Comment by ovflowd** [comment]
### :information_source: Updates on Development

Once we start working on the new Components, we want to start writing them in compliance with React Client Components (aka add a use 'client'; on top of each Component).

That’s because we aim to switch to the [Next.js app router](https://nextjs.org/docs) with the Website Redesign. During the Component implementation, pretty much nothing changes besides the above statement. It is, of course, recommended to check how React Client Components work. In our case, all the Design System Components are Client Components.

For Layouts, the same applies. As we will continue to use Markdown-defined Layouts. This means that we’re not using nested-Layouts (aka dedicated `layout.tsx` file for each directory level. This is mostly due to our subdirectories being dynamically generated and the root-level being a Dynamic Language code.

````
## Current `pages/_app.tsx` and `pages/_document.tsx` (contents of `_document.tsx` pretty much can be ignored as Next.js injects Head/Body, etc automatically)
- `app/layout.tsx`

## Base path of our dynamic generation
- `app/[...path]/`
  - `page.tsx` // this is what the current `pages/[...path].tsx` does
````

---

#### Note on Metadata/Frontmatter

It's also important to mention that Page Metadata still comes from a [Markdown Frontmatter](https://daily-dev-tips.com/posts/what-exactly-is-frontmatter/) file. We will primarily utilize [Next.js App Router Metadata Generation](https://nextjs.org/docs/app/api-reference/functions/generate-metadata). Technically, this means we will read a Markdown file twice for each route. However, for Metadata Generation, we can simply use graymatter, which is all we need. One of the ideas is to modify our Next.js Blog Data Generation Script to encompass all pages. This way, we can conveniently record all-page Metadata in a single JSON file. This would eliminate the need to read the Markdown file twice or even access the Markdown Frontmatter dynamically, as it is already stored within the static build-time JSON file.

Of course, this comes with the side-effect of having a large JSON file that the application reads. I assume that since we're removing these duplicated props from Next.js itself, the difference in bundle size would be similar.

---

#### Note on Dynamic Page Generation

We will mostly be using these functions to accomplish similar results:
- [generate-static-params](https://nextjs.org/docs/app/api-reference/functions/generate-static-params)
- [static-site-generation-getstaticprops](https://nextjs.org/docs/app/building-your-application/upgrading/app-router-migration#static-site-generation-getstaticprops)
- [next-mdx-remote](https://github.com/hashicorp/next-mdx-remote#react-server-components-rsc--nextjs-app-directory-support)

---

#### Note on other interesting Docs

The App Router introduces different semantics and new features. For example, it includes support for [OG Image Generation](https://nextjs.org/docs/app/api-reference/functions/generate-image-metadata) and built-in Analytics. If you're interested in the major changes, we recommend reading **Next.js's Upgrade Migration**.

- For Layout and Component Development, this won't affect newcomers and existing collaborators significantly, as the changes required for components are minimal. Therefore, we recommend using [Next.js's Client Component Guide](https://nextjs.org/docs/app/building-your-application/rendering/client-components) when creating new components.
- When we start migrating the pages, I'll be working on the updated Next Dynamic Page Generation Engine (`next.dynamic.mjs`) to support both the current `pages/` and the new `app/` router seamlessly. This transition should be clean and allow us to easily remove legacy website content when necessary.
- This transition should be transparent for the average contributor.

**Comment by ovflowd** [comment]
#### Hey y'all 👋 just a quick friendly announcement that we've started already working on the new Component implementation!

Please feel free to grab any Issue with the `website redesign` label on this repository!

**Reply by shaikahmadnawaz** [reply]
Okay, thanks for the announcement @ovflowd ✨

**Comment by patrick-mcdougle** [comment]
While this website is being built is the old website going to have up to date information? For example, Node 20 flipped into LTS and the page needs to be updated.

**Reply by ovflowd** [reply]
> While this website is being built is the old website going to have up to date information? For example, Node 20 flipped into LTS and the page needs to be updated.

Which page needs to be updated? The Website Redesign replaces the "old website" but the contents are the same.

**Reply by patrick-mcdougle** [reply]
https://nodejs.dev/en/about/releases/

As I understand v20 is now "active LTS".

**Reply by ovflowd** [reply]
nodejs.dev is not a maintained website. This repository is nodejs.org, not nodejs.dev...

**Reply by patrick-mcdougle** [reply]
Right, so we don't get an updated website until the redesign is done? Which was my original question.

**Reply by ovflowd** [reply]
Right. I've disabled nodejs.dev for the moment to avoid confusion. But yes, until the current ongoing work (that is steadily happening) is done. Also, the content of the website (except the guides section) is up to date.

But a few pages will indeed change.

**Reply by ovflowd** [reply]
We've added the remainder of the content of nodejs.dev to the main website!

**Comment by ovflowd** [comment]
## :information_source: Updates on Development

Here are more updates on what's going on with the Node.js Website Redesign:

- We finished the migration to Next.js's App Router. This allows us to use the greatest and latest features of Next.js and brings numerous improvements on bundling, bootstrap, separation of client-side/server-side (reduce of bundle-sizes, what is rendered on the server and what on the client, etc.)
- Continued improvements of the legacy Website (removal of the `legacyMain.js` file, adoption of Shikiji as a syntax highlighter, other minor styles, and bug fixes while we're working on the redesign
- We're nearing the end of the First Phase of the Website Redesign, which is the actual development of the React Components. Very few remaining now. (I believe 2-3 issues still need to be created for the 2-3 remaining components, but after that, we can start working on the layouts)
- We've adopted a new Internationalization System (`next-intl`) and are working closely with the maintainers of `next-intl` and providing them feedback (so we can enable Turbopack and other neat features)
  - `next-intl` has great integration with Next.js but can also run standalone (without Next.js)
- We've adopted Sentry for reporting Exceptions and Monitoring our Application, and we're working closely with the Sentry team to bring optimal configuration to our setup. Right now we're trying to tweak a few performance and bundling optimisations for Sentry

We are super excited with the direction of the work and even more as we're nearing the next step of the Website Redesign with implementing the Layouts on Storybook.

#### What would be the remaining steps?

- Create the Layouts and use them on Storybook with "demo" of how each page would look like
- Create a branch with adoption of the new Layouts over the pages
  - This might require some extra logic to be implemented and maybe a few hooks here and there for the optimal coverage of this
- Deploy Live the new Layouts (Website Redesign)
- Remove the old Layouts, old styles and remnant codebase of the Legacy Website

**Reply by AugustinMauroy** [reply]
> Create a branch with adoption of the new Layouts over the pages

I'm not a fan of branching (bad memory of porting nodejs.dev components).

**Reply by ovflowd** [reply]
> I'm not a fan of branching (bad memory of porting nodejs.dev components).

Well, but it is necessary here. We cannot incrementally adopt the layouts. It needs to be a one-go.

I believe either this will be one PR to be honest; Since the creation of each layout is a PR (not a branch -- pointing to main) and then the adoption of layouts is pretty much the swap on the `withLayout.tsx` file, and hopefully just some extra logic if needed.

**Reply by AugustinMauroy** [reply]
For branch I allow with you.
But I thinks logic of `withLayout.tsx` is wrong approche. we should using routing of Next.js
like that:
```
app/
├─ [locale]/
│  ├─ layout.tsx
│  ├─ page.tsx
│  ├─ about/
│  │  ├─ [...slug]/
│  │  │  ├─ page.tsx
│  ├─ learn/
│  │  ├─ [[...slug]]/
│  │  │  ├─ page.tsx
...
```
 So with that layout doesn't depend of front matter but it's depend of URL.
 Olson nextjs will genrate one executable for `/learn`, `/`, `/about` ... so in theory the server response should be faster because it should not calculate the full catch

**Reply by ovflowd** [reply]
> For branch I allow with you.
> 
> But I thinks logic of `withLayout.tsx` is wrong approche. we should using routing of Next.js
> 
> like that:
> 
> ```
> 
> app/
> 
> ├─ [locale]/
> 
> │  ├─ layout.tsx
> 
> │  ├─ page.tsx
> 
> │  ├─ about/
> 
> │  │  ├─ [...slug]/
> 
> │  │  │  ├─ page.tsx
> 
> │  ├─ learn/
> 
> │  │  ├─ [[...slug]]/
> 
> │  │  │  ├─ page.tsx
> 
> ...
> 
> ```
> 
>  So with that layout doesn't depend of front matter but it's depend of URL.
> 
>  Olson nextjs will genrate one executable for `/learn`, `/`, `/about` ... so in theory the server response should be faster because it should not calculate the full catch

Im not going to argue with you for the 5th time about why your approach doesn't work and why it also does not bring any benefit to the table.

**Reply by ovflowd** [reply]
After reflecting, I'll write here why the approach above is not suitable and why it is technically challenging.

- The Node.js Website allows any Markdown page to have a different Layout of their choice. This is part of the Frontmatter of every page, and it is done so in a way that allows newcomers and existing contributors to easily add new pages (including root-level pages, i.e.,/learn) or change layouts of existing pages. It also allows us to be unbound on Next.js's path-based router. It allows us to adopt any Layout as desired on any page of any level of the path-based router.
  - The approach above would nullify that, disallowing us to choose layouts dynamically based on each page's unique requirements.
- The approach above is "path-bounded"; it requires that if we create, for example, a new section of the website (i.e., https://nodejs.org/en/a-new-category/) to duplicate the existing logic for our dynamic router. That's a lot of unnecessary code duplication, hindering the developer experience and unnecessarily complicating our codebase.
- Our dynamic router works by reading a glob of the whole `/pages` file structure and creating a dictionary of allowed paths that our website supports. This is completely dynamic (Except for a few rules (https://github.com/nodejs/nodejs.org/blob/main/next.constants.mjs#L110-L112) (Note that the pagination of the Blog will change in the future, so these rules will be removed, but it's just an example) and allows us to provide to Next.js on our top-level `[[...path]]` dynamic segment all the paths that are supported by our website.
  - With the approach above, we would need to add an extra logic fragment to our dynamic router and be able to explicitly split what each catch-all segment of the example above should handle. This creates more unnecessary logic that needs to run on the Application's bootstrap, Including on ISR (Incremental Static Rendering), as we need to calculate the allowed paths for every request. (That's the nature of our ISR approach. Of course, we could remove ISR and build all pages during regular builds as we do with `npm run deploy` for our static exports because then all pages would be created at one go, but we would be worsening the Developer Experience just because the cold-boot of every page is longer)
  - In other words, we would make the whole logic behind our dynamic router fragmented and fragile. Not to mention it would increase, possibly a hundreds of milliseconds if not 1-2 seconds) the fresh-rendering of every page on ISR. Remember, ISR rendering is serverless and requires a cold-boot of the Application. Every extra logic on top of that just worsens the User-Experience and the performance of the Website.
- There are issues with using catch-all segments with other sub-paths such as `/blog.`
  - The `/blog` subpath would require at least two dynamic segments, one for `/[locale]/blog/[category]/page.tsx` responsible for handling the index page and categories and `/[locale]/blog/[category]/[slug]/page.tsx` accountable for managing each blog post
  - While the example above feels more structured, it's another example of all the extra code we need to maintain and handle by destructuring (removing) our utmost top-level catch-all segment.
- While the claim of the approach above is of an easier/more structured "app" folder, we don't care about the "app" folder; it has only one purpose: Handling the Requests of our Pages for all the content that exists within our "pages" folder. We don't need a per-layout structured folder as, in practical terms, it brings no benefits.
- There are no performance improvements. The approach above claims to have performance gains, But there are not. Using the `WithLayout` HoC simply wraps each page with a Layout.
  - The logic within `withLayout` is straightforward. At worst, O(1), since we're just passing a prop and rendering a Component with the name of the given prop (https://github.com/nodejs/nodejs.org/blob/main/components/withLayout.tsx). There is no performance loss, nor a more extensive client-side or per-page bundle, as the `WithLayout` component is a React Server Component. The HoC will never become part of the client bundle. The processing on the server side is diminished; the current approach has no performance bottlenecks.
  - But the new approach creates even more dynamic segments that Next.js needs to evaluate and handle at build and runtime; there would be more logic and complexity, and we would lose many of the freedoms of the current Website.

These are the reasons why the given approach is not suitable for us. I have explained a few times shortly before, and I know you have the best intentions, Augustin. But I cannot have you questioning every inch of the technical design of the Website because you feel like it. Imagine if everyone asked about Fastify's design. It is an annoyance for a maintainer to be questioned needlessly about specific architectural designs. Each of us has different opinions of how X should be done, And it is valid that maintainers are held accountable. Still, having someone from the team needlessly question when we are already overwhelmed with many other tasks is annoying.

I understand you have opinions and ideas, and they are always welcome, but the timing is just not the best. I have already explained the current architecture and why we are doing things this way a few times; please spare me the already little time I have with your questions.

**Reply by rojete** [reply]
@ovflowd this person is as "17 years old junior dev" according to their profile, so it's a natural situation IMO. 

I wonder if the problem comes from poor selection of team members?

**Reply by ovflowd** [reply]
I understand that Augustin is young and lacks experience in the field. But he conquered his place in the team. I'm not going to lie; sometimes, there are rough patches... But as you said, he's only 17 years old, and I can't expect him to act as a versed person in the industry.

His addition to the team was reasoned over a long time. I'm confident and supportive of all the team members of the Node.js Website and that they have their own journeys and learning trajectories.

**Comment by ValeriiRuchko** [comment]
Can you please tell when the new design of the site will be shipped to the audience? 👉👈

**Reply by ovflowd** [reply]
No. Because there's no date.

**Reply by ovflowd** [reply]
But I'd recommend you keep a tab on the PRs we're opening :) it will become a bit obvious when we're closer to the finish line. Also I'll be updating this thread once we have more updates!

**Reply by ValeriiRuchko** [reply]
Thank you for your answer and I wish you good luck :D

**Comment by ovflowd** [comment]
## :information_source: Updates on Development

Hey y'all 👋 it's been a while since I've posted updates here! So I wanted to come up with shiny updates!

### Component Development is complete

The development of all essential Components needed for the Node.js Website Redesign is complete. From now on, only development of React Components specific to certain Layouts will be done; And these will be done within the Pull Requests for each Layout.

### We're implementing the Website Redesign Layouts

For a while already, the Node.js Team has been implementing the Website Redesign Layouts within the core of the Node.js Website. Remember that these layouts are behind a feature flag, and won't be visible until explicitly requested. You can run locally the Node.js Website with the new Layouts enabled with `npm run serve:redesign`

#### What is the state of development of Layouts?

- [x] Base Layout (Base Styling)
  - This is a transitive/wrapper layout, and not directly used by pages
- [x] Article Layout (Article-specific Styling)
  - This is a transitive/wrapper layout, and not directly used by pages
  - Used by Learn and About Layouts
- [x] Centered Layout (Pages with Centered Content)
  - Used by Home Layout and Error/NotFound Pages
- [x] Blog Layout (For Blog Index and Category Pages)
- [x] Post Layout (For Blog Posts and Posts in general)
- [x] Learn Layout (For Learn Pages)
- [x] Docs Layout (For Deprecated Node.js Guides)
- [x] Content Layout
  - This is a transitive/wrapper layout, and not directly used by pages
  - Used by Docs and Post Layouts
- [x] About Layout (For About and Get Involved Pages)
- [x] Home Layout (For the Home Page)
- [x] Default Layout (For any other Page)
- [x] Download Layout (For Download Pages)

### What's next and remaining in order to get the Website Redesign released?

The following tasks still need to be done before we are able to release the Node.js Website with the Redesign:

- [x] Finish the missing Layouts
- [x] Migrate deprecated Node.js Guides to Learn Pages
- [x] Remove deprecated content
- [x] Enable the Website Redesign with a feature flag to test traffic and look for Errors
- [x] Remove legacy Website Pages, Content, functions, styles and codebase
- [x] Remove Website Redesign Feature Flag
- [x] Remove Legacy Website i18n keys
- [x] Enable Node.js Website on Crowdin so that translations can start to happen again

---

Thank you all for following the Node.js Website redesign updates!

**Comment by ovflowd** [comment]
It's funny how quickly a discussion might need to be updated! One month passed, and we've finished the Downloads Layouts. We're working on the bug fixes and UI improvements! We're live on https://beta-node-js-org.vercel.app and getting things ready to release to production ✨

**Reply by nazarepiedady** [reply]
@ovflowd, you and the team are doing an excellent, I can not wait 🤤 to start to translate the content and help to build the website (right after I polish my React and Next.js knowledge 😅).

**Reply by Igorgro** [reply]
Website is down now, returning 404

**Reply by AugustinMauroy** [reply]
> Website is down now, returning 404

Which link is death ?

**Reply by ovflowd** [reply]
This links doesn't matter anymore as we concluded the redesign. You can simply access https://nodejs.org
